package NYT.NRpcProxy.NProto;

import "yt/core/rpc/rpc.proto";
import "yt/core/misc/guid.proto";

////////////////////////////////////////////////////////////////////////////////

message TCredentialsExt
{
    extend NRpc.NProto.TRequestHeader
    {
        optional TCredentialsExt credentials_ext = 110;
    }

    optional string userip = 1;
    optional string token = 2;
    optional string domain = 3;
    optional string sessionid = 4;
    optional string sslsessionid = 5;
}

////////////////////////////////////////////////////////////////////////////////

enum ERowsetKind
{
    UNVERSIONED = 0;
    VERSIONED = 1;
}

// Specifies how to decode rowset data passed via attachments.
message TRowsetDescriptor
{
    // Currently, there is only one version. This field is reserved for future changes.
    optional int32 wire_format_version = 1 [default = 1];

    // Specifies, what kind of rows are encoded.
    optional ERowsetKind rowset_kind = 2 [default = UNVERSIONED];

    // Column id-to-name mapping.
    repeated string column_names = 4;

    // Column id-to-type mapping.
    // Currently unused, because both unversioned and versioned rowsets are self-describing.
    repeated int32 column_types = 5;
}

////////////////////////////////////////////////////////////////////////////////

message TReqGetNode
{
    required string path = 1;
}

message TRspGetNode
{
    required bytes data = 1;
}

message TReqLookupRows
{
    required string path = 1;

    repeated string columns = 2;
    optional uint64 timestamp = 3 [default = 0x3fffffffffffff01];
    optional bool keep_missing_rows = 4 [default = true];

    required TRowsetDescriptor rowset_descriptor = 200;
}

message TRspLookupRows
{
    required TRowsetDescriptor rowset_descriptor = 200;
}

message TReqVersionedLookupRows
{
    required string path = 1;

    repeated string columns = 2;
    optional uint64 timestamp = 3 [default = 0x3fffffffffffff01];
    optional bool keep_missing_rows = 4 [default = true];

    required TRowsetDescriptor rowset_descriptor = 200;
}

message TRspVersionedLookupRows
{
    required TRowsetDescriptor rowset_descriptor = 200;
}

////////////////////////////////////////////////////////////////////////////////
